onnxruntime/core/providers/xnnpack/nn/resize.cc

```diff
@@ -0,0 +1,257 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/xnnpack/nn/resize.h"
+
+#include <utility>
+
+#include "core/common/common.h"
+#include "core/framework/op_kernel.h"
+#include "core/optimizer/initializer.h"
+
+namespace onnxruntime {
+namespace xnnpack {
+
+bool Resize::IsOnnxNodeSupported(const NodeUnit& node_unit,
+                                 const GraphViewer&) {
+  bool supported = false;
+  do {
+    // Resize has 1-4 input.
+    const auto& inputs = node_unit.Inputs();
+    const auto& x_arg = inputs[0].node_arg;
+
+    const auto* x_type = x_arg.TypeAsProto();
+    if (x_type == nullptr ||
+        (x_type->tensor_type().elem_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT &&
+         x_type->tensor_type().elem_type() != ONNX_NAMESPACE::TensorProto_DataType_UINT8 &&
+         x_type->tensor_type().elem_type() != ONNX_NAMESPACE::TensorProto_DataType_INT8)) {
+      break;
+    }
+    const auto* x_shape = x_arg.Shape();
+    //'bilinear' == 2-D input or 4-D input with outermost 2 scales as 1 or
+    // 4-D input with outermost and innermost scales as 1
+    // but we just support 4-d tensor for now
+    if (!x_shape || x_shape->dim_size() != 4 || x_shape->dim(1).dim_value() <= 0) {
+      break;
+    }
+    const auto* output_shape = node_unit.Outputs()[0].node_arg.Shape();
+    bool length_resized_compatible_pytorch_half_pixel = true;
+    if (output_shape->dim(2).dim_value() <= 1 || output_shape->dim(1).dim_value() <= 1) {
+      length_resized_compatible_pytorch_half_pixel = false;
+    }
```

Comment:
I am assuming onnx will infere its shape when build the model.

It's a little bit tricky to decide if we support `pytorch_half_pixel`. On the one hand, it's quite usual to have `pytorch_half_pixel` since most of model was converted from pytorch. We'll miss a huge chance to optimize it with XNNPACK. On the other hand,  xnnpack's compatiability on this Op is depends on output shape.

If we claimed capability on this Op but failed in Compute, is it a bug?

I have tested this op with a model converted from pytorch, it works fine.

Reply:
Yes, the onnx spec says:
```
if coordinate_transformation_mode is "half_pixel",
x_original = (x_resized + 0.5) / scale - 0.5

if coordinate_transformation_mode is "pytorch_half_pixel",
x_original = length_resized > 1 ? (x_resized + 0.5) / scale - 0.5 : 0
```

Reply:
`Channel` to be known is a hard requirement for xnnpack kernel while it's not for the H and W. A dynamic (H,W) for 4-dim tensor only cost a little to calculate output shape.